ucm2/Allwinner/A64/PinePhone/HiFi.conf

```diff
@@ -0,0 +1,144 @@
+SectionVerb {
+	EnableSequence [
+		cset "name='Headphone Playback Switch' off"
+		cset "name='Headphone Source Playback Route' DAC"
+		cset "name='Line In Playback Switch' off"
+		cset "name='Line Out Playback Switch' off"
+		cset "name='Line Out Source Playback Route' Mono Differential"
+		cset "name='Mic1 Playback Switch' off"
+		cset "name='Mic2 Playback Switch' off"
+		cset "name='AIF1 DA0 Playback Volume' 160"
+		cset "name='AIF3 ADC Source Capture Route' None"
+		cset "name='AIF2 DAC Source Playback Route' None"
+		cset "name='DAC Playback Switch' on"
+		cset "name='DAC Playback Volume' 160"
+		cset "name='ADC Digital DAC Playback Switch' off"
+		cset "name='AIF1 Slot 0 Digital DAC Playback Switch' on"
+		cset "name='AIF2 Digital DAC Playback Switch' off"
+		cset "name='DAC Reversed Playback Switch' off"
+		cset "name='Earpiece Playback Switch' off"
+		cset "name='Earpiece Source Playback Route' DACL"
+
+		cset "name='Line In Capture Switch' off"
+		cset "name='Mic1 Capture Switch' off"
+		cset "name='Mic1 Boost Volume' 7"
+		cset "name='Mic2 Capture Switch' off"
+		cset "name='Mic2 Boost Volume' 7"
+		cset "name='Mixer Capture Switch' off"
+		cset "name='Mixer Reversed Capture Switch' off"
+		cset "name='ADC Capture Volume' 160"
+		cset "name='ADC Gain Capture Volume' 7"
+		cset "name='AIF1 AD0 Capture Volume' 160"
+		cset "name='AIF1 Data Digital ADC Capture Switch' on"
+		cset "name='AIF2 ADC Mixer ADC Capture Switch' off"
+		cset "name='AIF2 ADC Mixer AIF1 DA0 Capture Switch' off"
+		cset "name='AIF2 ADC Mixer AIF2 DAC Rev Capture Switch' off"
+	]
+
+	DisableSequence [
+	]
+
+	Value {
+	}
+}
+
+SectionDevice."Speaker" {
+	Comment "Internal speaker"
+	EnableSequence [
+		cset "name='AIF1 DA0 Stereo Playback Route' Mix Mono"
+		cset "name='Line Out Playback Switch' on"
+		cset "name='Line Out Playback Volume' 100%"
+	]
+
+	DisableSequence [
+		cset "name='Line Out Playback Switch' off"
+	]
+
+	Value {
+		PlaybackVolume "Line Out Playback Volume"
+		PlaybackSwitch "Line Out Playback Switch"
+		PlaybackChannels 2
+		PlaybackPriority 300
+		PlaybackPCM "hw:${CardId},0"
+	}
+}
+SectionDevice."Earpiece" {
+	Comment "Internal Earpiece"
+	EnableSequence [
+		cset "name='AIF1 DA0 Stereo Playback Route' Mix Mono"
```

Comment:
The "Speaker" and "Earpiece" sections don't conflict with "Headphones". So enabling both "Speaker" and "Headphones" would give you mix mono audio to the headphones. I think a nice solution here is, at the HiFi profile level:
 * Enable "DAC Playback Switch"
 * Enable "DAC Reversed Playback Switch"
 * Set "AIF1 DA0 Stereo Playback Route" to "Stereo"
 * Set "Headphone Source Playback Route" to "DAC"
 * Set "Earpiece Source Playback Route" to "Left Mixer"

Then, the headphones will always be stereo, because they bypass the analog mixer. At the same time, since both DAC channels will go into both sides of the analog mixer, the earpiece and speaker will get mix mono.

Reply:
If "DAC Reversed Playback Switch" is enabled on the HiFi profile, and an XMPP video/audio chat is active, then AFAICT the output would be fed into the input, as if it came from the mic.

AFAICT, an XMPP video/audio chat wouldn't want to use the Voice Call profile, since the Voice Call profile sets "PlaybackRate 8000".

As an alternative, should we just set that "Speaker" and "Earpiece" each conflict with "Headphones". I think this matches user expectations too.

Reply:
"DAC Playback Switch" and "DAC Reversed Playback Switch" control the exact same path from the DACs to the analog Output Mixer, except that the left and right channels are swapped. They are the two pairs of "m"s overlaying the rightward-pointing arrows just to the right of the green "DACL" and "DACR" boxes in this diagram: https://xnux.eu/devices/feature/audio-controls.svg.

The "Reversed" here refers to the stereo channels, not to the direction.

(The only loopback paths in the codec are controlled by the "Mixer (Reversed) Capture Switch"; see the leftward-pointing arrows at the bottom of the diagram. Those are always off in both profiles.)

Reply:
Thanks for the explanation, that makes sense.

I like this approach because it means our routing can be only: routes don't change, only switches do. That makes this config easier to understand.

I have updated this PR accordingly: https://github.com/alsa-project/alsa-ucm-conf/pull/134/commits/fb6e94983e69b902c66306214a4cfab3beabb7fa.